[FLINK-29405] Fix unstable test InputFormatCacheLoaderTest

SmirAlex opened the pull request:
<!--
*Thank you very much for contributing to Apache Flink - we are happy that you want to help us improve Flink. To help the community review your contribution in the best possible way, please go through the checklist below, which will get the contribution into a shape in which it can be best reviewed.*

*Please understand that we do not do this to make contributions to Flink a hassle. In order to uphold a high standard of quality for code contributions, while at the same time managing a large number of contributions, we need contributors to prepare the contributions well, and give reviewers enough contextual information for the review. Please also understand that contributions that do not follow this guide will take longer to review and thus typically be picked up with lower priority by the community.*

## Contribution Checklist

  - Make sure that the pull request corresponds to a [JIRA issue](https://issues.apache.org/jira/projects/FLINK/issues). Exceptions are made for typos in JavaDoc or documentation files, which need no JIRA issue.
  
  - Name the pull request in the form "[FLINK-XXXX] [component] Title of the pull request", where *FLINK-XXXX* should be replaced by the actual issue number. Skip *component* if you are unsure about which is the best component.
  Typo fixes that have no associated JIRA issue should be named following this pattern: `[hotfix] [docs] Fix typo in event time introduction` or `[hotfix] [javadocs] Expand JavaDoc for PuncuatedWatermarkGenerator`.

  - Fill out the template below to describe the changes contributed by the pull request. That will give reviewers the context they need to do the review.
  
  - Make sure that the change passes the automated tests, i.e., `mvn clean verify` passes. You can set up Azure Pipelines CI to do that following [this guide](https://cwiki.apache.org/confluence/display/FLINK/Azure+Pipelines#AzurePipelines-Tutorial:SettingupAzurePipelinesforaforkoftheFlinkrepository).

  - Each pull request should address only one issue, not mix up code from multiple issues.
  
  - Each commit in the pull request has a meaningful commit message (including the JIRA id)

  - Once all items of the checklist are addressed, remove the above text and this checklist, leaving only the filled out template below.


**(The sections below can be removed for hotfixes of typos)**
-->

## What is the purpose of the change
Fix unstable test InputFormatCacheLoaderTest. 
The problem was caused by race condition when we interrupt Full cache reload thread. There is the following logic: when main reload thread is being interrupted, it also interrupts underlying threads, that read from InputSplits (InputSplitCacheLoadTask). Previously, main thread didn't wait for the end of underlying threads, so test sometimes failed, if test thread progressed faster than InputSplitCacheLoadTask threads.
Note, that previously if we put, for instance, `@RepeatedTest(1000)` on this test, it would be failed almost every time. Now, with  `@RepeatedTest` it never fails (but I think, it's unnecessary to have this annotation on continuing basis).


## Brief change log

  - *Fixed unstable tests*
  - *Fixed `InputFormatCacheLoader`, so now after exception it waits until `InputSplitCacheLoadTasks` finish*

## Verifying this change

This change is already covered by existing tests

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (no)
  - If yes, how is the feature documented? (not applicable)

HuangXingBo commented:
cc @PatrickRen

MartijnVisser commented:
@zentol Any thoughts?

zentol reviewed (changes requested):
Waiting forever in production code is super sketchy and should virtually never be done.

The PR is also lacking and sort of problem analysis and explanation for how this fixes the issue.

SmirAlex commented:
> Waiting forever in production code is super sketchy and should virtually never be done.
> 
> The PR is also lacking a sort of problem analysis and explanation for how this fixes the issue.

Hi @zentol, I added timeout on wait after interrupt and updated PR description to explain the problem and proposed solution more precisely. Can you check the latest commit, please?

XComp commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/inputformat/InputFormatCacheLoader.java:
```diff
@@ -107,7 +109,11 @@ protected void reloadCache() throws Exception {
             if (cacheLoadTaskService != null) {
+                // if main cache reload thread encountered an exception,
+                // it interrupts underlying InputSplitCacheLoadTasks threads
                 cacheLoadTaskService.shutdownNow();
```
Out of curiosity: Was it a deliberate decision to rely on the commonForkPool (see [ReloadTriggerContext:52](https://github.com/apache/flink/blob/10a6f41fa12284f15af55c359c0c0954800f02de/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/ReloadTriggerContext.java#L52) if `numSplits == 1` and create a dedicated thread pool if we have more than 1 split? I'm just wondering whether aligning both would be an option by relying on `CompletableFuture.runAsync` here as well and, that way, avoiding to wait for the shutdown.

I couldn't find any reasoning to do that other than readability/less code.

SmirAlex replied:
To be honest, during development I didn't think about relying on ForkJoinPool. As I understand, it's very useful when we have many little tasks, that also can be created in already running tasks. But in our case we have fixed amount ( = number of splits) of long running tasks, that won't create other subtasks. So simple fixed thread pool looked as a pretty straightforward way to implement cache loading. Plus, if we will use commonForkPool, there can be a situation when we will utilize all threads in this pool (our tasks are long-running), and other tasks that relies on commonForkPool with be starving, which is undesirable behavior, as I understand. Correct me, if I'm wrong, please

zentol replied:
I'm not sure what the question is.

We certainly shouldn't use the common fork pool. I can see what caused that decision to be made, but I'd say the wrong conclusion was drawn. reloadCache should've explictly been an async operation. That would also make it obvious when looking at the class where concurrency actually exists.

If this operation is so heavy that you don't want to call it in the calling thread then it probably shouldn't run in the common pool.

Not a huge fan of repeatedly creating a new executor service btw; I'd rather create it once for a consistent resource profile. As a side-effect it would then also be trivial to make this call async.

XComp replied:
yeah, I was just curious about your opinion on the usage of common pool. It felt like implementing it asynchronously would have worked here. I just struggled to think of a better way to enable thread loading based on the number of `InputSplits` to optimize the resource utilization per cache loading cycle.

SmirAlex replied:
In my latest commit I changed `CacheLoader`, so now it returns future directly + it doesn't rely on common ForkJoinPool. Have a look, please.

SmirAlex commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/inputformat/InputFormatCacheLoader.java:
```diff
@@ -39,12 +39,14 @@
 public class InputFormatCacheLoader extends CacheLoader {
     private static final long serialVersionUID = 1L;
     private static final Logger LOG = LoggerFactory.getLogger(InputFormatCacheLoader.class);
+    private static final long TIMEOUT_AFTER_INTERRUPT = 10; // 10 sec
```
Agree, fixed

SmirAlex commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/inputformat/InputFormatCacheLoader.java:
```diff
@@ -107,7 +109,11 @@ protected void reloadCache() throws Exception {
+                // if main cache reload thread encountered an exception,
+                // it interrupts underlying InputSplitCacheLoadTasks threads
                 cacheLoadTaskService.shutdownNow();
+                // timeout 10 sec should definitely be enough to wait for finish after interrupt
```
Agree, fixed

SmirAlex commented:
Hi @XComp, thanks for the review! I answered on your comments and added a commit with fix. Have a look when you will be available, please.

XComp reviewed (approved):
But just on a side note (independent of this PR's change): I might be wrong but it feels like we could have implemented `reloadCache` in an async fashion returning a `CompletableFuture`. Because when investigating the call hierachy: We're calling `reloadCache` in a `Runnable#run` which we execute in `ReloadTriggerContext#triggerReload` anyway asynchronously. I'm just mentioning it because I'm curious what your view is on that.

XComp commented on flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/functions/table/fullcache/inputformat/InputFormatCacheLoaderTest.java:
```diff
@@ -149,19 +155,33 @@ void testCloseAndInterruptDuringReload() throws Exception {
+    @Test
+    void testCloseDuringReload() throws Exception {
+        AtomicInteger recordsCounter = new AtomicInteger(0);
+        int totalRecords = TestCacheLoader.DATA.size() + 1; // 1 key with 2 records
```
```suggestion
        int totalRecords = TestCacheLoader.DATA.values().stream().mapToInt(Collection::size).sum();
```
nit: you could use this call. That way you could remove the comment which easily becomes wrong if someone changes `TestCacheLoader.DATA`. WDYT?

SmirAlex replied:
Agree, fixed

SmirAlex commented on flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/functions/table/fullcache/inputformat/InputFormatCacheLoaderTest.java:
```diff
@@ -149,19 +155,33 @@ void testCloseAndInterruptDuringReload() throws Exception {
-        assertThatNoException().isThrownBy(future::get); // wait for the end
+        latch.countDown();
+        future.get(); // wait for the end
+        executorService.shutdown();
```
But do we need to wait here? Currently there are 2 ways to stop cache loading: via `CacheLoader#close()` method and via interrupting a thread. This one test only first way of stopping. Actually, we could limit ourselves to the first way only, but I decided to also implement interrupt handling for completeness and, probably, more fast reaction on closing.

SmirAlex replied:
Agree, now all tests on `InputFormatCacheLoader` uses try-with-resources, so all threads used for cache loading should be closed after test passes

XComp commented on flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/functions/table/fullcache/inputformat/InputFormatCacheLoaderTest.java:
```diff
@@ -149,19 +155,33 @@ void testCloseAndInterruptDuringReload() throws Exception {
+                ThrowingRunnable.unchecked(
+                        () -> {
+                            recordsCounter.incrementAndGet();
+                            latch.await();
```
Doing that leads to a deadlock. Even just adding `Thread.sleep` shortly before closing the cache loader makes the test run into a deadlock because [CacheLoader:101](https://github.com/apache/flink/blob/340b100f2de5e0d90ba475aa8a00e359a61442ce/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/CacheLoader.java#L101) acquires the `CacheLoader.reloadLock` and initiates the reloading of the cache synchronously while having the lock acquired. The data reload blocks due to the `CountDownLatch` we use in the test. The subsequent call of close triggers the shutdown of the `InputSplitCacheLoadTask` instances through `InputSplitCacheLoadTask#stopRunning`. But the shutdown of those tasks will never be done because the data loading is still blocked and the tasks are stuck in the [while loop](https://github.com/apache/flink/blob/340b100f2de5e0d90ba475aa8a00e359a61442ce/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/inputformat/InputSplitCacheLoadTask.java#L71). [InputFormatCacheLoader#close](https://github.com/apache/flink/blob/a024a366f73f822bb4fd35db737ac2b8177f6b25/flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/inputformat/InputFormatCacheLoader.java#L129) will continue with calling `CacheLoader#close` which tries to acquire the lock that's already acquired by the thread that executes `InputFormatCacheLoader` and we end up in a deadlock.

My suspicion is that the test as it is pushed right now is passing because we're calling close before the data reload is initiated. AFAIU, that's not what we want to test, is it? Please correct me if I'm wrong here.

XComp reviewed (commented):
Thanks for sharing your thoughts here, @SmirAlex. And sorry for the slow pace and multiple iterations. I still need to familiarize myself with that part of the codebase. Anyway, I had a bit of a closer look at the test code once more. I'm looking forward to your view on the comments on `testCloseDuringReload`.

SmirAlex commented:
Hi @XComp @zentol, thanks for your careful review! I took into account all your comments and created a commit with following changes: 

- Now `CacheLoader` returns future directly in method `CacheLoader#reloadAsync()`. As a consequence `CacheLoader` has direct control over the thread which executes cache reload operation.
- I no more rely on common ForkJoinPool. `CacheLoader` creates `Executors.newSingleThreadExecutor()` in `#open` method that is responsible for the whole reload operation.
- Now `CacheLoader` can directly shutdown its executor in `#close`, so I decided to remove close mechanism based on boolean variable checking and keep only the one based on thread interruption. Therefore logic became more clear and straightforward. So now there is only one test on `CacheLoader#close`
- Now `InputFormatCacheLoader` executes one load task in main reload thread (in `Cacheloader#reloadExecutor`), so we are saving on creating additional thread for reading from input split.
- Improved logging on `CacheLoader#close` a little bit

Please take a look on the latest commit when you will be available, looking forward for your review!

SmirAlex commented on flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/functions/table/fullcache/inputformat/InputFormatCacheLoaderTest.java:
```diff
@@ -131,15 +133,19 @@ void testExceptionDuringReload() throws Exception {
+                        // wait should be interrupted if everything works ok
+                        if (!recordsProcessingLatch.await(5, TimeUnit.SECONDS)) {
+                            throw new RuntimeException("timeout");
+                        }
```
Agree, I reused this code part in my latest commit with changed `CacheLoader#close`

SmirAlex commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/inputformat/InputFormatCacheLoader.java:
```diff
@@ -107,7 +109,11 @@ protected void reloadCache() throws Exception {
             if (cacheLoadTaskService != null) {
+                // if main cache reload thread encountered an exception,
+                // it interrupts underlying InputSplitCacheLoadTasks threads
                 cacheLoadTaskService.shutdownNow();
```
In our case we need to call `ExecutorService#shutdown` immediately, so `ExecutorUtils#gracefulShutdown` doesn't fit well. But I improved logging in case of error

XComp reviewed (commented):
I didn't go through eveyrthing, yet. I just want to propose something since you started to change method signatures a bit more.

About the responsibilities how I see them:
* `LookupFullCache` is the composite class for combining the `CacheLoader` and the `CacheReloadTrigger` through the `ReloadTriggerContext`
* `ReloadTriggerContext` provides an async call to trigger the reload but also some utility methods for providing processing or event time (where it's not clear to me why this is connected with the reload)
* `CacheLoader` is in charge of loading the data into memory (if possible concurrently).
* `CacheReloadTrigger` provides different strategies to trigger new reloads.

About the different executors:
* The `CacheReloadTrigger` utilize a `SingleThreadScheduledExecutor` which triggers `ReloadTriggerContext::reload` subsequently. If the loading takes longer, subsequently triggered calls pile up. Here, I'm wondering whether that's what we want. :thinking: 
* `CacheLoader` utilizes a `SingleThreadExecutor` in `CacheLoader#reloadExecutor` which is kind of the "main" thread for reloading the data. It triggers `CacheLoader#updateCache` with `CacheLoader#reloadLock` being acquired. `(InputFormat)CacheLoader#updateCache` is implemented synchronously. The data is loaded concurrently if possible using a `FixedThreadPool`.

My proposal is now to reduce the number of used threadpools: Instead of having a `SingleThreadExecutor` and a `FixedThreadPool`, couldn't we come up with a custom `ThreadPoolExecutor` where we specify the minimum number of threads being 1 and the maximum being the number of cores. That would free the `CacheLoader` from starting and shutting down thread pools by moving its ownership from `CacheLoader` to `LookupFullCache` calling it the `cacheLoadingThreadPool` (or similar). Additionally, the `ScheduledThreadPool` currently living in the `CacheReloadTrigger` implementations could move into `LookupFullCache` as well calling it something like `cacheLoadSchedulingThreadPool`. `LookupFullCache` would be in charge of managing all cache loading-related threads. Additionally, it would manage the current execution through `CompletableFutures` (one for triggering the reload and one for executing the reload. Triggering a reload would require cancelling the current future (if it's not completed, yet) or ignoring the trigger if we want a reload to finish before triggering a new one.

`CacheLoader#updateCache` would become `CacheLoader#updateCacheAsync(ExecutorService)` returning a `CompletableFuture` that completes as soon as all subtasks are completed. `CacheLoader#reloadAsync` would return this `CompletableFuture` instead of creating its own future. The lifecycle (as already explained in the previous paragraph) would be managed by `LookupFullCache`. The benefit would be that we wouldn't have to deal interrupts in `CacheLoader`. 

I see the following benefits:
* `ReloadtriggerContext` becomes obsolete (one has to clarify what the event time and processing time functions are for, though).
* `CacheLoader#awaitFirstLoad` becomes obsolete as well. We can verify the completion of the cache loading in `LookupFullCache` through the `CompletableFuture` instances.
* `CacheReloadTrigger` can focus on the strategy implementation without worrying about instantiating threads. This is duplicated code right now in `PeriodicCacheReloadTrigger` and `TimedCacheReloadTrigger`.

I might miss something here. I'm curious what you think. I probably got carried away a bit by your proposal introducing async calls. :innocent: I totally understand if you argue that it's way too much out-of-scope for this issue and we actually want to focus on fixing the test instability. In that case, I would do another round of review on your current proposal. But I'm happy to help you if you think that my proposal is reasonable. Or we create a follow-up Jira issue to tackle that.

XComp commented:
After some offline chat with @zentol about it, I came to the conclusion that it's really not the right location for such a big change. We should focus on stabilizing the test here. I moved my proposal into FLINK-30354. I will do another review of this PR on Monday.

SmirAlex commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/CacheLoader.java:
```diff
@@ -59,11 +65,11 @@ public abstract class CacheLoader extends AbstractRichFunction implements Runnab
+        reloadExecutor = Executors.newSingleThreadExecutor();
     }
 
     public void open(CacheMetricGroup cacheMetricGroup) {
```
Agree, fixed. Btw you are right, naming was initially came from `LookupCache#open`

SmirAlex commented:
> After some offline chat with @zentol about it, I came to the conclusion that it's really not the right location for such a big change. We should focus on stabilizing the test here. I moved my proposal into [FLINK-30354](https://issues.apache.org/jira/browse/FLINK-30354). I will do another review of this PR on Monday.

Yea, I think this MR is already a little bit blowed up, so such big changes should be in separate ticket. About your proposal: I like the idea of reusing the same thread pool, and especially, clarifying the logic of `CacheLoader`, so it won't deal with interrupts. But I'm not sure about changing `CacheReloadTrigger`s and `CacheReloadTrigger.Context`, because it touches already public API (although added relatively recently). BTW during the [FLIP-221 ](https://cwiki.apache.org/confluence/display/FLINK/FLIP-221:+Abstraction+for+lookup+source+cache+and+metric) development I discussed the design with @PatrickRen and @wuchong. So I think it would be good if they also take a look at your proposal in [FLINK-30354](https://issues.apache.org/jira/browse/FLINK-30354)

XComp commented on flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/functions/table/fullcache/inputformat/InputFormatCacheLoaderTest.java:
```diff
@@ -123,46 +126,47 @@ void testExceptionDuringReload() throws Exception {
+        }
+        // try-with-resources calls #close, which should wait for the end of reload
+        assertThat(future.isDone()).isTrue();
+        assertThat(metricGroup.loadCounter.getCount()).isEqualTo(1);
```
That's kind of unexpected: We're closing the loader by interrupting the threads executing `InputSplitCacheLoadTask`. Is this actually where we want to consider the cache loading to be a success? :thinking:

SmirAlex replied:
Hmm, you are right. Fixed

XComp commented on flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/functions/table/fullcache/inputformat/InputFormatCacheLoaderTest.java:
```diff
@@ -123,46 +126,47 @@ void testExceptionDuringReload() throws Exception {
-        assertThat(sleepCounter).hasValueLessThan(totalSleepCount);
-        assertThat(metricGroup.numLoadFailuresCounter.getCount()).isEqualTo(0);
+    static Stream<Arguments> numSplits() {
+        return Stream.of(Arguments.of(1), Arguments.of(2));
```
AFAIU, `numSplits` being `1` means that there is no extra fixed-size thread pool created in the `InputFormatCacheLoader`, but for `2` we expect additional threads to be created. Is this considered an implementation detail or shall we test this feature through some `@VisibleForTesting` utility method in `InputFormatCacheLoader` that returns the threadpool (which would be `null` in case of `numSplits=1`). :thinking:

SmirAlex replied:
I think it can be implementation detail, especially considering that [FLINK-30354](https://issues.apache.org/jira/browse/FLINK-30354) can significantly change internal logic

XComp reviewed (commented):
Sorry for the late reply. I didn't manage to review it last week. I was on sick leave.
I'm still not a big fan of how we use the interruption and how we have thread pools in different locations. But we cover this in FLINK-30354. And I agree with you @SmirAlex that we should include @PatrickRen in this discussion as well.

Please find my comments below.

PS: @PatrickRen it would be good to get your opinion on this change as well

SmirAlex commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/inputformat/InputFormatCacheLoader.java:
```diff
@@ -71,58 +72,58 @@ public void open(Configuration parameters) throws Exception {
+            CompletableFuture<?> otherTasksFuture = null;
+            if (!cacheLoadTasks.isEmpty()) {
+                cacheLoadTaskService = Executors.newFixedThreadPool(concurrencyLevel - 1);
+                ExecutorService finalExecutor = cacheLoadTaskService;
```
Otherwise it's impossible to refer to `cacheLoadTaskService` in subsequent lambda because this variable is not final

XComp replied:
ah true, I overlooked that one. 8)

SmirAlex commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/LookupFullCache.java:
```diff
@@ -57,7 +57,7 @@ public synchronized void open(CacheMetricGroup metricGroup) {
+        cacheLoader.initializeMetrics(metricGroup);
     }
 
     public synchronized void open(Configuration parameters) throws Exception {
```
I agree that 2 `open` methods is a problem, not only because of naming, but because of ambiguity in the order of calling these methods, and non-compliance with the contract of `LookupCache#open`, which should fully open the cache in one call. So I decided to keep just one `open` method and pass `Configuration` in different method. As a drawback we now have redundant state in cache, but I think it's a good compromise in our circumstances.

SmirAlex commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/LookupFullCache.java:
```diff
@@ -109,7 +109,7 @@ public long size() {
     public void close() throws Exception {
-        reloadTrigger.close(); // firstly try to interrupt reload thread
+        reloadTrigger.close();
         cacheLoader.close();
```
Added comments

SmirAlex commented on flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/functions/table/fullcache/inputformat/FullCacheTestInputFormat.java:
```diff
@@ -86,20 +92,21 @@ public FullCacheTestInputFormat(
         this.rowConverter = rowConverter;
-        this.deltaNumSplits = 0;
+        this.numSplits = DEFAULT_NUM_SPLITS;
+        this.deltaNumSplits = DEFAULT_DELTA_NUM_SPLITS;
```
Fixed

SmirAlex commented on flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/functions/table/fullcache/inputformat/InputFormatCacheLoaderTest.java:
```diff
@@ -77,43 +77,46 @@ void checkCounter() {
+            assertCacheContent(cache);
+            run(cacheLoader);
+            // new instance of cache after reload
+            assertThat(cacheLoader.getCache()).isNotSameAs(cache);
```
Thanks, fixed

SmirAlex commented on flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/functions/table/fullcache/inputformat/InputFormatCacheLoaderTest.java:
```diff
@@ -77,43 +77,46 @@ void checkCounter() {
+            run(cacheLoader);
+            // new instance of cache after reload
+            assertThat(cacheLoader.getCache()).isNotSameAs(cache);
+            cacheLoader.close();
```
Fixed

SmirAlex commented on flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/functions/table/fullcache/inputformat/InputFormatCacheLoaderTest.java:
```diff
@@ -123,46 +126,47 @@ void testExceptionDuringReload() throws Exception {
+                createCacheLoader(numSplits, DEFAULT_DELTA_NUM_SPLITS, reloadAction)) {
+            cacheLoader.initializeMetrics(metricGroup);
+            future = cacheLoader.reloadAsync();
+            reloadLatch.await(5000, TimeUnit.MILLISECONDS);
```
Fixed

SmirAlex commented:
Hi @XComp, I'm sorry I've been gone so long. I had a vacation to pass university exams and then got ill 😬. I answered all your comments and made a fix in recent commit (force-pushed only because of rebase). Have a look please, when you will have time.

XComp commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/LookupFullCache.java:
```diff
@@ -109,7 +116,9 @@ public long size() {
     @Override
     public void close() throws Exception {
-        reloadTrigger.close(); // firstly try to interrupt reload thread
+        // in default triggers shutdowns scheduled thread pool used for periodic triggers of reloads
```
```suggestion
        // stops scheduled thread pool that's responsible for scheduling cache updates
```
Just as a proposal

SmirAlex replied:
fixed

SmirAlex commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/inputformat/InputFormatCacheLoader.java:
```diff
@@ -71,56 +72,56 @@ public void open(Configuration parameters) throws Exception {
+                        .collect(Collectors.toCollection(ArrayDeque::new));
+        // run first task and create concurrencyLevel - 1 threads to run remaining tasks
         ExecutorService cacheLoadTaskService = null;
+        boolean wasInterrupted;
```
fixed

SmirAlex commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/CachingLookupFunction.java:
```diff
@@ -107,12 +107,12 @@ public void open(FunctionContext context) throws Exception {
-        // Initialize cache and the delegating function
-        cache.open(cacheMetricGroup);
         if (cache instanceof LookupFullCache) {
             // TODO add Configuration into FunctionContext
```
Yes, I will create one, thanks for advice

SmirAlex replied:
https://issues.apache.org/jira/browse/FLINK-30778

SmirAlex commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/CachingLookupFunction.java:
```diff
@@ -107,12 +107,12 @@ public void open(FunctionContext context) throws Exception {
         if (cache instanceof LookupFullCache) {
             // TODO add Configuration into FunctionContext
-            ((LookupFullCache) cache).open(new Configuration());
+            ((LookupFullCache) cache).setParameters(new Configuration());
```
I added this block for future change. But now it's redundant, agree. Fixed

SmirAlex commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/LookupFullCache.java:
```diff
@@ -109,7 +116,9 @@ public long size() {
-        reloadTrigger.close(); // firstly try to interrupt reload thread
+        // in default triggers shutdowns scheduled thread pool used for periodic triggers of reloads
+        reloadTrigger.close();
+        // shutdowns the reload thread and interrupts active reload task, if present
```
fixed

SmirAlex commented:
@XComp @PatrickRen At this time I fixed all comments that was by this MR. If there will be no more remarks, existing commits can be squashed into one and merged to master. After that I will create the same MR, but if Flink 1.16.

XComp commented on flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/functions/table/fullcache/inputformat/FullCacheTestInputFormat.java:
```diff
@@ -41,18 +41,22 @@
 
     public static final AtomicInteger OPEN_CLOSED_COUNTER = new AtomicInteger(0);
-    private static final int DEFAULT_NUM_SPLITS = 2;
+    public static final int DEFAULT_NUM_SPLITS = 2;
+    public static final int DEFAULT_DELTA_NUM_SPLITS = 0;
```
```suggestion
    static final int DEFAULT_NUM_SPLITS = 2;
    static final int DEFAULT_DELTA_NUM_SPLITS = 0;
```

XComp replied:
tbh, it appears that we could even make these values `private`. The only time, they are used is in `InputFormatCacheLoaderTest` which doesn't seem to rely on these values but just uses them as "some" input value. Or is there a specific reason why we need the default values as input values? :thinking:

SmirAlex replied:
Agree, changed to `private`

XComp commented on flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/functions/table/fullcache/inputformat/FullCacheTestInputFormat.java:
```diff
@@ -41,18 +41,22 @@
 
-/** TestInputFormat that reads data from (2 + delta) splits which share the same {@code queue}. */
+/**
+ * TestInputFormat that reads data from (initNum + delta) splits which share the same {@code queue}.
```
```suggestion
 * TestInputFormat that reads data from ({@link #numSplits} + {@link #deltaNumSplits}) splits which share the same {@code queue}.
```
What is `initNum`? I guess, you were referring to the newly added field `numSplits` analogously to the field `deltaNumSplits`?

XComp commented on flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/functions/table/fullcache/inputformat/InputFormatCacheLoaderTest.java:
```diff
@@ -123,46 +127,48 @@ void testExceptionDuringReload() throws Exception {
+            future = cacheLoader.reloadAsync();
+            reloadLatch.await();
+        }
+        // try-with-resources calls #close, which should wait for the end of reload
+        assertThat(future.isDone()).isTrue();
```
```suggestion
        // try-with-resources calls #close which will interrupt any running threads, which should wait for the end of reload
        assertThat(future.isDone()).as("The reload future should still complete successfully indicating that the reload was intentionally stopped without an error.").isTrue();
```
Is this correct? At least that's how I understand this test. :thinking: Feel free to change/revert

SmirAlex replied:
Fixed

XComp commented on flink-table/flink-table-runtime/src/test/java/org/apache/flink/table/runtime/functions/table/fullcache/inputformat/InputFormatCacheLoaderTest.java:
```diff
@@ -178,12 +184,16 @@ private void assertCacheContent(Map<RowData, Collection<RowData>> actual) {
                         assertThat(rows).containsExactlyInAnyOrderElementsOf(actual.get(key)));
     }
 
+    private void run(CacheLoader cacheLoader) {
```
```suggestion
    private void reloadSynchronously(CacheLoader cacheLoader) {
```
nit: that would be a better description of what this method is doing.

SmirAlex replied:
Fixed

XComp reviewed (approved):
Thanks for your quick response. I did a final pass over the entire PR and had a few minor suggestions. Otherwise, the change looks reasonable to fix the test instability. Just for the record: I had a more fundamental concern on how the thread pools are initialized. We agreed that such a refactoring would be out-of-scope for FLINK-29405. Therefore, I moved my suggestions into FLINK-30354 as a follow-up to improve testability/maintainability of the code.

I approve this PR under the premise that there's a final round of review being done by @PatrickRen since I am not that familiar with that part of the code. AFAIK, @PatrickRen will be back from the Lunar Year holidays end of this week or beginning of next week.

Thanks again, @SmirAlex :-)

SmirAlex commented on flink-table/flink-table-runtime/src/main/java/org/apache/flink/table/runtime/functions/table/lookup/fullcache/ReloadTriggerContext.java:
```diff
@@ -26,11 +26,11 @@
 
-    public ReloadTriggerContext(Runnable reloadTask, Consumer<Throwable> reloadFailCallback) {
-        this.reloadTask = reloadTask;
+    public ReloadTriggerContext(CacheLoader cacheLoader, Consumer<Throwable> reloadFailCallback) {
```
Agree, changed to functional interface

XComp commented:
Thanks @SmirAlex. May you squash and rebase the PR. This way, we could already get it final hoping that @PatrickRen finds time to have a pass till tomorrow. If not, I would still go ahead and merge the PR tomorrow before feature freeze to make this test instability go away at least. Based on how much you trust your PR, you could also create a 1.16 backport already to have CI run over night. ...up to you (as it might double your amount of work if @PatrickRen still has change requests).

PatrickRen reviewed (approved):
Thanks for the patch @SmirAlex and quite sorry for my late response. LGTM for stabilizing the CI. 

@XComp raised some quite good ideas in FLINK-30354 about improving the overall implement of full caching part. I think we could continue the discussion there. And big thanks to @XComp for reviewing the PR!

SmirAlex commented:
@XComp Big thanks for the review, your comments were very helpful! @PatrickRen Thanks for merging a MR! And sorry for not responding, I was on a vacation without laptop last week, so, unfortunately, couldn't merge it by myself.
